codec-http/src/main/java/io/netty/handler/codec/http/DefaultHttpHeaders.java

```diff
@@ -94,26 +96,42 @@ public CharSequence convertObject(Object value) {
             return seq;
         }
 
+        private static final class ValidateValueProcessor implements ByteVisitor {
+            private int state;
+            @Override
+            public boolean process(byte value) throws Exception {
+                state = validateValueChar(state, (char) value);
+                return true;
+            }
+
+            public int state() {
+                return state;
+            }
+        }
+
         private static void validateValue(AsciiString seq) {
-            int state = 0;
-            // Start looping through each of the character
-            final int start = seq.arrayOffset();
-            final int end = start + seq.length();
-            final byte[] array = seq.array();
-            for (int index = start; index < end; index++) {
-                state = validateValueChar(seq, state, (char) (array[index] & 0xFF));
+            ValidateValueProcessor processor = new ValidateValueProcessor();
+            try {
+                seq.forEachByte(processor);
+            } catch (Throwable t) {
+                PlatformDependent.throwException(
+                        new IllegalArgumentException("an illegal header value was detected: " + seq, t));
             }
 
-            if (state != 0) {
+            if (processor.state() != 0) {
                 throw new IllegalArgumentException("a header value must not end with '\\r' or '\\n':" + seq);
             }
         }
 
         private static void validateValue(CharSequence seq) {
             int state = 0;
-            // Start looping through each of the character
-            for (int index = 0; index < seq.length(); index++) {
-                state = validateValueChar(seq, state, seq.charAt(index));
+            try {
+                // Start looping through each of the character
+                for (int index = 0; index < seq.length(); index++) {
+                    state = validateValueChar(state, seq.charAt(index));
+                }
+            } catch (IllegalArgumentException e) {
+                throw new IllegalArgumentException("an illegal header value was detected: " + seq, e);
```

Comment:
Instead of creating two exception for an invalid header, could we just pass `seq` to `validateValueChar()` so that it produces proper exception message and we save an instantiation of an exception, which is expensive?